### PR TITLE
Enable AoT for net8 in Release

### PR DESCRIPTION
**Pull request by Arlodotexe**
Enables Native AoT functionality when running under .NET 8 in preparation for the WinAppSdk 1.6 release with NativeAoT support.  
 
Prerequisite PR: https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/190

### CODE REVIEW

**Comment by Arlodotexe** [comment]
This is a Draft PR until this remaining error can be fixed.
```
 error APPX1101: Payload contains two or more files with the same destination path 'CommunityToolkit.App.WinAppSdk.exe'.
```

Full error:
```
       "C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\CommunityToolkit.App.WinAppSdk.csproj" (Clean;Build target) (1:7) ->
       (_ComputeAppxPackagePayload target) ->
         C:\Users\arlog\.nuget\packages\microsoft.windowsappsdk\1.5.240311000\buildTransitive\Microsoft.Build.Msix.Packaging.targets(1515,5): error APPX1101: Payload contains two or more files with the same destination path 'CommunityToolkit.App.WinAppSdk.exe'. Source files:  [C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winap 
       psdk\CommunityToolkit.App.WinAppSdk.csproj]
       C:\Users\arlog\.nuget\packages\microsoft.windowsappsdk\1.5.240311000\buildTransitive\Microsoft.Build.Msix.Packaging.targets(1515,5): error APPX1101: C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\obj\x64\Release\net8.0-windows10.0.22621.0\win10-x64\apphost.exe [C:\Users\arlog\source\repos\Labs-Windows\tooling\Pr 
       ojectHeads\AllComponents\winappsdk\CommunityToolkit.App.WinAppSdk.csproj]
       C:\Users\arlog\.nuget\packages\microsoft.windowsappsdk\1.5.240311000\buildTransitive\Microsoft.Build.Msix.Packaging.targets(1515,5): error APPX1101: C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\bin\x64\Release\net8.0-windows10.0.22621.0\win10-x64\native\CommunityToolkit.App.WinAppSdk.exe [C:\Users\arlog\source 
       \repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\CommunityToolkit.App.WinAppSdk.csproj]
```

The above error is the same for both this repo and in Labs.

**Comment by michael-hawker** [comment]
> This is a Draft PR until this remaining error can be fixed.
> 
> ```
>  error APPX1101: Payload contains two or more files with the same destination path 'CommunityToolkit.App.WinAppSdk.exe'.
> ```
> 
> Full error:
> 
> ```
>        "C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\CommunityToolkit.App.WinAppSdk.csproj" (Clean;Build target) (1:7) ->
>        (_ComputeAppxPackagePayload target) ->
>          C:\Users\arlog\.nuget\packages\microsoft.windowsappsdk\1.5.240311000\buildTransitive\Microsoft.Build.Msix.Packaging.targets(1515,5): error APPX1101: Payload contains two or more files with the same destination path 'CommunityToolkit.App.WinAppSdk.exe'. Source files:  [C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winap 
>        psdk\CommunityToolkit.App.WinAppSdk.csproj]
>        C:\Users\arlog\.nuget\packages\microsoft.windowsappsdk\1.5.240311000\buildTransitive\Microsoft.Build.Msix.Packaging.targets(1515,5): error APPX1101: C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\obj\x64\Release\net8.0-windows10.0.22621.0\win10-x64\apphost.exe [C:\Users\arlog\source\repos\Labs-Windows\tooling\Pr 
>        ojectHeads\AllComponents\winappsdk\CommunityToolkit.App.WinAppSdk.csproj]
>        C:\Users\arlog\.nuget\packages\microsoft.windowsappsdk\1.5.240311000\buildTransitive\Microsoft.Build.Msix.Packaging.targets(1515,5): error APPX1101: C:\Users\arlog\source\repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\bin\x64\Release\net8.0-windows10.0.22621.0\win10-x64\native\CommunityToolkit.App.WinAppSdk.exe [C:\Users\arlog\source 
>        \repos\Labs-Windows\tooling\ProjectHeads\AllComponents\winappsdk\CommunityToolkit.App.WinAppSdk.csproj]
> ```
> 
> The above error is the same for both this repo and in Labs.

@Arlodotexe heard back from the platform team, this is a known issue and expected with 1.5. A fix should be coming in 1.6, so we can start testing that later as we see experimental/preview releases.

For now, let's just turn on the AoT compatible bit/annotations and get that checked-in.

File: components/Collections/src/AdvancedCollectionView/AdvancedCollectionView.cs [hunk 1]
```diff
@@ -13,6 +13,9 @@ namespace CommunityToolkit.WinUI.Collections;
 /// <summary>
 /// A collection view implementation that supports filtering, sorting and incremental loading
 /// </summary>
+#if NET8_0_OR_GREATER
+[System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("Item sorting uses reflection to get property types and may not be AOT compatible.")]
```
Comment (jp-weber) [inline components/Collections/src/AdvancedCollectionView/AdvancedCollectionView.cs:17]: Shouldn't the item sorting uses work without problems with AOT? At least with UWP with .NET native there are no problems, or are there changes in the implementation?

Comment (Arlodotexe) [reply components/Collections/src/AdvancedCollectionView/AdvancedCollectionView.cs:17]: [[docs](https://learn.microsoft.com/en-us/windows/uwp/dotnet-native/net-native-and-compilation#just-in-time-compilation)] UWP with .NET Native uses the default runtime directive for libraries: 
```
<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
  <Application>
    <Assembly Name="*Application*" Dynamic="Required All" />
  </Application>
</Directives>
```
> This enables all the types, as well as all their members, in all the assemblies in your app package for reflection and dynamic invocation. However, it does not enable reflection or dynamic activation of types in .NET Framework Class Library assemblies. In many cases, this is adequate.

For .NET 8 and NativeAOT, they use these attributes to bubble up warnings or errors to the compiler from underlying libraries to signal when AoT might not work as expected. The code change here is a simple suppression of one of these errors.

**Comment by Arlodotexe** [comment]
Here's the error message from the CI:
```
ExtensionsComponent.Tests.FrameworkElementExtensionsTests.FrameworkElementExtension_RelativeAncestor_FreeParentBaseline_FrameworkElementExtensionsTests_Test 
```

This seems to be a transient error, so I'll rerun the ci and we'll address it directly if it reproduces consistently.

File: components/Collections/src/IncrementalLoadingCollection/IncrementalLoadingCollection.cs [hunk 2]
```diff
@@ -130,6 +130,11 @@ private set
     /// An <see cref="Action"/> that is called if an error occurs during data retrieval.
     /// </param>
     /// <seealso cref="IIncrementalSource{TSource}"/>
+#if NET8_0_OR_GREATER
+    [System.Diagnostics.CodeAnalysis.RequiresUnreferencedCode("This constructor is not AOT compatible and will be removed in a future version.")]
+#endif
+    [EditorBrowsable(EditorBrowsableState.Never)]
+    [Obsolete("This constructor is not AOT compatible and will be removed in a future version.")]
```
Comment (michael-hawker) [inline components/Collections/src/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:137]: Can we provide more here or a remark/example of how to convert the use of an existing use of this constructor to a valid one?

Comment (Arlodotexe) [inline components/Collections/src/IncrementalLoadingCollection/IncrementalLoadingCollection.cs:137]:
```suggestion
    [Obsolete("This constructor is not AOT compatible and will be removed in a future version. Use the other constructor and provide a source collection to use.")]
```

File: components/Collections/src/AdvancedCollectionView/AdvancedCollectionView.cs (hunk 1 above)
Comment (jp-weber) [reply components/Collections/src/AdvancedCollectionView/AdvancedCollectionView.cs:17]: Ah :) Thank you for the explanation.

**Review by michael-hawker — APPROVED** [review]

**Pull request merged**